src/NuGet.Core/NuGet.Commands/ContentFiles/ContentFileUtils.cs

```diff
@@ -0,0 +1,286 @@
+                    );
+
+                groups.AddRange(contentFilesGroupsWithSameFramework);
+            }
+
+            return groups;
+        }
+
+        /// <summary>
+        /// Apply build actions from the nuspec to items from the contentFiles folder.
+        /// </summary>
+        internal static List<LockFileItem> GetContentFileGroup(
+            NuGetFramework framework,
+            NuspecReader nuspec,
+            List<ContentItemGroup> contentFileGroups)
+        {
+            var results = new List<LockFileItem>(contentFileGroups.Count);
+
+            // Read the contentFiles section of the nuspec
+            // Reverse the entries so that the top entry has priority
+            var nuspecContentFiles = nuspec.GetContentFiles().Reverse().ToList();
+
+            // Initialize mappings
+            var entryMappings = new Dictionary<string, List<ContentFilesEntry>>(StringComparer.OrdinalIgnoreCase);
+            var languageMappings = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var group in contentFileGroups)
+            {
+                var codeLanguage = group.Properties[ManagedCodeConventions.PropertyNames.CodeLanguage] as string;
+
+                foreach (var item in group.Items)
+                {
+                    if (!entryMappings.ContainsKey(item.Path))
+                    {
+                        entryMappings.Add(item.Path, new List<ContentFilesEntry>());
+                        languageMappings.Add(item.Path, codeLanguage);
+                    }
+                }
+            }
+
+            // Virtual root for file globbing
+            var rootDirectory = new VirtualFileInfo(VirtualFileProvider.RootDir, isDirectory: true);
+
+            // Apply all nuspec property mappings to the files returned by content model
+            foreach (var filesEntry in nuspecContentFiles)
+            {
+                // this is validated in the nuspec reader
+                Debug.Assert(!string.IsNullOrEmpty(filesEntry.Include), "invalid contentFiles entry");
+
+                // Create a filesystem matcher for globbing patterns
+                var matcher = new Matcher(StringComparison.OrdinalIgnoreCase);
+                matcher.AddInclude(filesEntry.Include);
+
+                if (filesEntry.Exclude != null)
+                {
+                    matcher.AddExclude(filesEntry.Exclude);
+                }
+
+                // Check each file against the patterns
+                foreach (var file in entryMappings.Keys)
+                {
+                    // Remove contentFiles/ from the string
+                    Debug.Assert(file.StartsWith("contentFiles/", StringComparison.OrdinalIgnoreCase),
+                        "invalid file path: " + file);
+                    var relativePath = file.Length > 13 ? file.Substring(13, file.Length - 13) : file;
```

Comment:
Perhaps, we should also assert that the length cannot be <= 13. And, handle the cases the file is not valid. Error or skip as approrpriate. So, relativePath should always be the substring and no need of the else